### ISSUE TITLE
Window maximum width and height issues on Linux Wayland

### ISSUE THREAD

**Issue by spinzed** [wailsapp/wails#2431]
### Description

There are 3 similar issues that I've found:
1. It is not possible to resize window beyond a certain width and height for some reason (although MaxWidth and MaxHeight are unset).
2. Maximizing the window will move the window to the top left of the screen but it won't take up the entire screen (similarly to the first bug).
3. When making the window fullscreen, the window itself will go fullscreen but the webview inside it won't and will center itself in the window.

### To Reproduce

For all:
1. Make sure you are on Wayland and the app is not ran under Xwayland.
2. Unset MaxWidth and MaxHeight or set them to a high value (close to the screen width and height)
3. Launch app

First bug:
- Try to resize manually and make the window as big as possible.
Second bug:
- Click maximize.
Third bug:
- Fullscreen the app.

### Expected behaviour

First bug: The app resizes accordingly without constraints.
Second bug: The app maximizes.
Third bug: The app goes to fullscreen.

### Screenshots

[Screencast from 2023-02-28 14-06-22.webm](https://user-images.githubusercontent.com/46136813/221863521-452f8709-e2df-421c-9838-5cf75f37eeff.webm)

### System Details

```shell
# System

OS           | Manjaro Linux
Version      | Unknown      
ID           | manjaro      
Go Version   | go1.19.4     
Platform     | linux        
Architecture | amd64        

# Wails

Version         | v2.3.1
Package Manager | pacman

# Dependencies

Dependency | Package Name | Status    | Version    
*docker    | docker       | Available | 1:23.0.1-1 
gcc        | gcc          | Installed | 12.2.1-2   
libgtk-3   | gtk3         | Installed | 1:3.24.36-1
libwebkit  | webkit2gtk   | Installed | 2.38.5-1   
npm        | npm          | Installed | 8.19.2-1   
pkg-config | pkgconf      | Installed | 1.8.0-1    
* - Optional Dependency
```

### Additional context

I have tried to reproduce the problem under Xwayland (launching the app with GDK_BACKEND=x11 and XDG_SESSION_TYPE=X11) but the bugs weren't present there.

I also successfully managed to reproduce the bugs on another machine running Arch with Gnome on Wayland.

Strangely, it also seems that the "phantom maximum width and height" depend on the GTK theme - for some themes they are a bit bigger and for some they are a bit smaller.

**Comment by leaanthony** [comment]
Sounds like an upstream bug. Maybe @Lyimmi or @tmclane have more knowledge

**Comment by Lyimmi** [comment]
Sounds wonderful, I'll take a look at it as soon as I have time this week.

**Comment by Lyimmi** [comment]
Hi, sorry it took me so long to look into this. I think I found the core of the issue.

Wails sets default min and max sizes for the window when they are not set in the app options (to the current display's resolution) and we pass the window geometry hints `GDK_HINT_MAX_SIZE | GDK_HINT_MIN_SIZE` to GTK as well, so GTK tries to respect the set window dimension limits. 

If I remove the hints in v2/internal/frontend/desktop/linux/window.c L:190 the problem goes away, but then we have no built in window size control via GTK.

As far as I can tell the default Wayland compositor counts the shadows/borders/etc in the window's width and height.

For a quick and dirty temporary fix I recommend to set the window max values in the app options to some arbitrary big number (maybe over a 4k displays resolution).

I'll look into how fix this properly.
By the way this is not a unique Wails problem, I found lots of other issues about this, including electron.

**Comment by leaanthony** [comment]
Thanks for looking into this @Lyimmi 🙏

**Comment by spinzed** [comment]
Thank you for checking this out! I've tried to look into this myself, but I am far too unfamiliar with GTK to do anything here, especially when taking into account those Wayland quirks.

Nevertheless, I'm grateful for your efforts and I'll provide you with any additional info if I find any. Cheers!

**Comment by alexaandru** [comment]
FWIW, I experience the same issue. Installed `wails` today, ran the sample app and then just hit "toggle maximize" key combo on the window, it only gets it partially maximized (top+left, leaving a gap on bottom+right).

<details>

<summary>Wails doctor</summary>

```
# Wails
Version         | v2.7.1
Package Manager | dnf   

# System
┌────────────────────────────────────────────────────────────────────────────────┐
| OS           | Fedora Linux                                                    |
| Version      | 39                                                              |
| ID           | fedora                                                          |
| Go Version   | go1.21.6                                                        |
| Platform     | linux                                                           |
| Architecture | amd64                                                           |
| CPU          | Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz                       |
| GPU 1        | GM107M [GeForce GTX 960M] (NVIDIA Corporation) - Driver: nvidia |
| GPU 2        | HD Graphics 530 (Intel Corporation) - Driver: i915              |
| Memory       | 15GB                                                            |
└────────────────────────────────────────────────────────────────────────────────┘

# Dependencies
┌────────────────────────────────────────────────────────┐
| Dependency | Package Name        | Status    | Version |
| *docker    | moby-engine         | Installed | 24.0.5  |
| gcc        | gcc-c++             | Installed | 13.2.1  |
| libgtk-3   | gtk3-devel          | Installed | 3.24.41 |
| libwebkit  | webkit2gtk4.0-devel | Installed | 2.42.3  |
| npm        | nodejs-npm          | Installed | 10.2.3  |
| pkg-config | pkgconf-pkg-config  | Installed | 1.9.5   |
| *upx       | upx                 | Installed | 4.2.2   |
└─────────────── * - Optional Dependency ────────────────┘
```

</details>


[Screencast from 2024-02-06 12-11-13.webm](https://github.com/wailsapp/wails/assets/85237/9f760d87-1f9c-40e7-8a2f-888ee0927bb2)

I use Gnome 45.3 on Wayland.

**Comment by lyimmi** [comment]
I think I found a way to fix this issue, but I have not seen this kind of solution anywhere, so I'm not exactly sure if it is a good solution.

Basically we could compare the window's initial size and the allocated size that contains the window decorators and shadows as well and add the difference to the window's max size values.

The comparison can be ignored on frameless windows because non decorated windows report the correct size on Wayland.

As far as I can tell the initial window size does not mater, the size of the decorators does not change just the window's size.

The solution still behaves badly on my portrait display, because for some reason Wayland returns the main display's resolution... So the max height is about half of the portrait display's height. At this time I can't figure out why.

What do you think @leaanthony, is it good enough for a PR?

It would look like this:

```C

void SetMinMaxSize(GtkWindow *window, int min_width, int min_height, int max_width, int max_height)
{
    GdkGeometry size;
    size.min_width = size.min_height = size.max_width = size.max_height = 0;

    GdkRectangle monitorSize = getCurrentMonitorGeometry(window);
    if (isNULLRectangle(monitorSize))
    {
        return;
    }
    int flags = GDK_HINT_MAX_SIZE | GDK_HINT_MIN_SIZE;
    size.max_height = (max_height == 0 ? monitorSize.height : max_height);
    size.max_width = (max_width == 0 ? monitorSize.width : max_width);
    size.min_height = min_height;
    size.min_width = min_width;

// wayland specific code start
# ifdef GDK_WINDOWING_WAYLAND
    char *gdkBackend = getenv("XDG_SESSION_TYPE");
    if (gdkBackend != NULL && strcmp(gdkBackend, "wayland") == 0 && gtk_window_get_decorated(window)) {
        int windowWidth, windowHeight;
        gtk_window_get_size(window, &windowWidth, &windowHeight);

        GtkAllocation windowAllocation;
        gtk_widget_get_allocation(GTK_WIDGET(window), &windowAllocation);

        // only for debug
        g_printf("window decorated: %d\n", gtk_window_get_decorated(window));
        g_printf("window alloc width:%d height:%d\n", windowAllocation.width, windowAllocation.height);
        g_printf("window window width:%d height:%d\n", windowWidth, windowHeight);
        g_printf("window & decorator diff width:%d height:%d\n", (windowAllocation.width-windowWidth), (windowAllocation.height-windowHeight));
        g_printf("window max width:%d height:%d\n", size.max_width, size.max_height);

        size.max_height = (windowAllocation.height-windowHeight)+size.max_height;
        size.max_width = (windowAllocation.width-windowWidth)+size.max_width;

        // only for debug
        g_printf("window new max width:%d height:%d\n", size.max_width, size.max_height);
    }
#endif
// wayland specific code end

    gtk_window_set_geometry_hints(window, NULL, &size, flags);
}
```

It prints:

```shell
window decorated: 1
window alloc width:1452 height:857
window window width:1400 height:768
window & decorator diff width:52 height:89
window max width:2560 height:1080
window new max width:2612 height:1169

```

**Comment by leaanthony** [comment]
@lyimmi - I can't believe this slipped through the net! This looks like a good solution! Are you still ok to open a PR for it?

**Comment by lyimmi** [comment]
@leaanthony - Sure, give me some time, I'm a little busy lately!

**Comment by lyimmi** [comment]
Hi, I think my PR solves this issue, please try out the PR!

@nathabonfim59, @alexaandru, @spinzed, @Morten-Renner

**Comment by alexaandru** [comment]
Hello!

Maximizing by dragging (mouse) works from the PR with one exception: I can "over-maximize" vertically (which I don't think it should, right?) but I cannot do that horizontally, which is what I'd expect. So I can maximize to the size of the screen and then go a little over at the bottom... ~about the size of the titlebar? Something in that range, as you can see in the attached video:

https://github.com/user-attachments/assets/a4fa6f52-d4c2-4bff-860a-cb25409eb86b

I have no idea how to test fullscreen though :-) On my window manager (Gnome 47/Wayland), `F11` key makes apps fullscreen, doesn't seem to do anything to the wails apps. I did try with both `Fullscreen: true` and `false`, no change. If you could tell me how to make the app fullscreen, I'd be happy to test that too.

Cheers!

**Comment by lyimmi** [comment]
I'm not too concerned that the window can be scaled beyond the display's boundaries.
On Ubuntu 24.04 Gnome I can't do that.
(I'm sure there are some distros that you can mirror the display if you drag it too far. 😄)

I don't know how to figure out, how big is the window's title, or if the window manager calculates the "top bar" as usable screen space, in a generic way. As far as I can tell most software sets the default min-max to MAX_INT anyway.
But is someone can help with that, that would be awesome.

So I'm okay with this behavior. @leaanthony what do you think?

For the fullscreen you could use the runtime to set the window to fullscreen, or just start it as fullscreen like so:

```golang

import (
    "github.com/wailsapp/wails/v2/pkg/options"
)

err := wails.Run(&options.App{
		Title:  "wails",
		Width:  1024,
		Height: 768,
                ...
		WindowStartState: options.Fullscreen,  <------ You need this
	        ...
	})
```

@alexaandru

**Comment by alexaandru** [comment]
Thank you, fullscreen works too, but that also works with the latest stable, not just the PR.

**Comment by lyimmi** [comment]
Yes, initial fullscreen works with master, but only once.
If you unfullscreen and fullscreen again it breaks for me.

**Comment by nathabonfim59** [comment]
@lyimmi 

Works like a charm on COSMIC Desktop (Wayland). Thank you very much!

**Comment by VirxEC** [comment]
This issue is still present in Wails v3.0.0-alpha.10

**Comment by leaanthony** [comment]
@VirxEC - please open a new ticket, fill in the required information and reference this ticket 👍

**Comment by VirxEC** [comment]
> Please note: No bug reports are currently being accepted for Wails v3

It says this when I go to report a bug, I'm not sure what to do
